### PR TITLE
Timeouts and multiple priv protocol in discovery/snmp.py

### PR DESCRIPTION
**Pull request by jjagodzinski**
Adding higher timeout (2 seconds) to every snmp function for properly
work with SNMPv3. Creating loop in snmp_command, snmp_bulk, snmp_walk
to get data from devices with different priv protocol.

### CODE REVIEW

File: src/ralph/discovery/snmp.py [hunk 1]
```diff
@@ -50,48 +57,59 @@ def user_data(auth, snmp_version, priv_protocol=cmdgen.usmDESPrivProtocol):
...
     transport = cmdgen.UdpTransportTarget((hostname, 161), attempts, timeout)
-    data = user_data(community, snmp_version, priv_protocol=priv_protocol)
-    gen = cmdgen.CommandGenerator()
-    error, status, index, vars = gen.getCmd(data, transport, oid)
+    for priv_protocol in PRIV_PROTOCOLS:
+        data = user_data(community, snmp_version, priv_protocol=priv_protocol)
+        gen = cmdgen.CommandGenerator()
+        error, status, index, vars = gen.getCmd(data, transport, oid)
+        if not error:
+            return vars
```
Comment (ghost) [inline src/ralph/discovery/snmp.py:68]: This `return` in loop is intentional? It's like "give me first nonerror result"?

File: src/ralph/discovery/snmp.py [hunk 2]
```diff
@@ -50,48 +57,59 @@ def user_data(auth, snmp_version, priv_protocol=cmdgen.usmDESPrivProtocol):
...
-    gen = cmdgen.CommandGenerator()
-    error, status, index, values = gen.nextCmd(data, transport, oid)
-    if not error:
-        return values
+    for priv_protocol in PRIV_PROTOCOLS:
+        data = user_data(community, snmp_version, priv_protocol=priv_protocol)
+        gen = cmdgen.CommandGenerator()
+        error, status, index, vars = gen.nextCmd(data, transport, oid)
+        if not error:
+            return vars
```
Comment (ghost) [inline src/ralph/discovery/snmp.py:81]: The same here.

File: src/ralph/discovery/snmp.py [hunk 3]
```diff
@@ -50,48 +57,59 @@ def user_data(auth, snmp_version, priv_protocol=cmdgen.usmDESPrivProtocol):
 
 
 def snmp_command(
-    hostname, community, oid, snmp_version='2c', timeout=1, attempts=3,
+    hostname, community, oid, snmp_version='2c', timeout=2, attempts=3,
     priv_protocol=cmdgen.usmDESPrivProtocol
```
Comment (ar4s) [inline src/ralph/discovery/snmp.py]: `priv_protocol` is override in `for` loop by local variable

File: src/ralph/discovery/snmp.py [hunk 4]
```diff
@@ -50,48 +57,59 @@ def user_data(auth, snmp_version, priv_protocol=cmdgen.usmDESPrivProtocol):
...
 ):
     transport = cmdgen.UdpTransportTarget((hostname, 161), attempts, timeout)
-    data = user_data(community, snmp_version, priv_protocol=priv_protocol)
-    gen = cmdgen.CommandGenerator()
-    error, status, index, vars = gen.getCmd(data, transport, oid)
+    for priv_protocol in PRIV_PROTOCOLS:
+        data = user_data(community, snmp_version, priv_protocol=priv_protocol)
+        gen = cmdgen.CommandGenerator()
```
Comment (ar4s) [inline src/ralph/discovery/snmp.py]: probably this line should be before `for` loop

Comment (vi4m) [inline src/ralph/discovery/snmp.py]: And that's what i call good review @ar4s and @xliiv - nicely done.

File: src/ralph/discovery/snmp.py (hunk 1 above)
Comment (jjagodzinski) [inline src/ralph/discovery/snmp.py:68]: Yes, it should work in this way.

**Comment by ar4s** [comment]
:ok:

**Comment by ghost** [comment]
:ok:

**Pull request merged**